### PR TITLE
4 default unsupervised discretization and grouping methods are wrong

### PR DESCRIPTION
**Pull request by folmos-at-orange**

### CODE REVIEW

File: khiops/core/internals/runner.py [hunk 1]
```diff
@@ -167,19 +168,15 @@ def _is_khiops_installed_in_a_conda_env():
...
-            if modl_ok and modl_cc_ok:
-                is_in_conda_env = True
-                break
+        bin_files = os.listdir(vendored_bin_path)
+        modl_ok = "MODL" in bin_files or "MODL.exe" in bin_files
+        modl_cc_ok = (
+            "MODL_Coclustering" in bin_files or "MODL_Coclustering.exe" in bin_files
+        )
+        if modl_ok and modl_cc_ok:
+            is_in_conda_env = True
```
Comment (popescu-v) [inline khiops/core/internals/runner.py:179]: Ok for `os.listdir` instead of `os.walk`, as the binaries are indeed expected to be directly inside `$CONDA_PREFIX/bin` and no recursive traversal of this directory is needed.

However, I would keep the `is_executable` test to make sure these files can actually be executed and are not some incompatible/random-format files. Thus, we prevent potential upstream bugs (at the packaging level).

**Comment by popescu-v** [comment]
> _ No description provided. _

What is the ticket associated to this PR?

**Review by popescu-v — CHANGES_REQUESTED** [review]
I do not see the relation between the name of this PR and the implemented changes. IMHO, we need to provide description and context (ticket) to this PR, or otherwise rename the PR to reflect the implemented changes.

File: khiops/core/internals/runner.py (hunk 1 above)
Comment (folmos-at-orange) [reply khiops/core/internals/runner.py:179]: I didn't realize I remove those checks. I'll reinstate them.

**Comment by folmos-at-orange** [comment]
In addition to the changes above I added the reporting of the stdout if not empty

**Comment by folmos-at-orange** [comment]
> > _ No description provided. _
> 
> What is the ticket associated to this PR?

It is on your right-side -->

**Comment by folmos-at-orange** [comment]
> I do not see the relation between the name of this PR and the implemented changes. IMHO, we need to provide description and context (ticket) to this PR, or otherwise rename the PR to reflect the implemented changes.

Just minor fixes that do not require another pull request. To go fast.

**Comment by popescu-v** [comment]
> > I do not see the relation between the name of this PR and the implemented changes. IMHO, we need to provide description and context (ticket) to this PR, or otherwise rename the PR to reflect the implemented changes.
> 
> Just minor fixes that do not require another pull request. To go fast.

I would just rename the PR.

**Comment by popescu-v** [comment]
> > > _ No description provided. _
> > 
> > 
> > What is the ticket associated to this PR?
> 
> It is on your right-side -->

Ok. Anyway this doesn't apply to the PR changes anyway. So, I would unlink the issue and just rename the PR.

File: khiops/extras/docker.py [hunk 2]
```diff
@@ -162,7 +162,7 @@ def _run(
         if trace:
             print(f"Docker runner delete job: {json_response}")
 
-        return return_code, stderr
+        return return_code, "", stderr
```
Comment (popescu-v) [inline khiops/extras/docker.py:165]: Why don't we return the proper `stdout` here, i.e. `response["response"]["output"]` (just like we do for `stderr`)?

If there is a reason, I would document it clearly, with a comment.

File: khiops/core/internals/runner.py [hunk 3]
```diff
@@ -616,51 +618,75 @@ def run(
...
             # Report the message if there were any
             if warning_msg:
                 warning_msg = (
                     "Khiops ended correctly but there were minor issues" + warning_msg
                 )
                 warnings.warn(warning_msg.rstrip(), stacklevel=4)
-        # If there were errors or fatal errors collect them and report
+        # If the execution was incorrect raise an exception reporting:
+        # - The errors and/or fatal errors in the log
+        # - The stdout if not empty
```
Comment (popescu-v) [inline khiops/core/internals/runner.py:659]: Perhaps add a comment that Khiops puts (some?) errors/warnings on stdout (so that it is clearer why we error on non-empty stdout)?
Or, to be even clearer, state what Khiops puts on stdout, on stderr, and what we want to capture here (why we issue warning if stdout is not empty, even though the return code is 0)?

**Review by popescu-v — CHANGES_REQUESTED** [review]

**Comment by folmos-at-orange** [comment]
> > > I do not see the relation between the name of this PR and the implemented changes. IMHO, we need to provide description and context (ticket) to this PR, or otherwise rename the PR to reflect the implemented changes.
> > 
> > 
> > Just minor fixes that do not require another pull request. To go fast.
> 
> I would just rename the PR.

I would not bother with it. What's important are the commits. The PR may dissapear tomorrow.

File: khiops/core/internals/runner.py (hunk 3 above)
Comment (folmos-at-orange) [reply khiops/core/internals/runner.py:659]: Ok.

File: khiops/extras/docker.py (hunk 2 above)
Comment (folmos-at-orange) [reply khiops/extras/docker.py:165]: Because the `"output"` in this case is the log.

File: khiops/core/internals/runner.py (hunk 3 above)
Comment (folmos-at-orange) [reply khiops/core/internals/runner.py:659]: Done. (I wrote a note at the beginning of the method).

File: khiops/core/internals/runner.py (hunk 1 above)
Comment (folmos-at-orange) [reply khiops/core/internals/runner.py:179]: In second thought, those checks were redundant. We execute `check_tools` afterwards which verifies exactly that.

**Comment by popescu-v** [comment]
> > > > I do not see the relation between the name of this PR and the implemented changes. IMHO, we need to provide description and context (ticket) to this PR, or otherwise rename the PR to reflect the implemented changes.
> > > 
> > > 
> > > Just minor fixes that do not require another pull request. To go fast.
> > 
> > 
> > I would just rename the PR.
> 
> I would not bother with it. What's important are the commits. The PR may dissapear tomorrow.

Ok.

File: khiops/extras/docker.py (hunk 2 above)
Comment (popescu-v) [reply khiops/extras/docker.py:165]: Ok; then, I would add a comment stating this before the return.

File: khiops/core/internals/runner.py (hunk 1 above)
Comment (popescu-v) [reply khiops/core/internals/runner.py:179]: Yeah, ok. There is a difference in behavior though:
- before, if either one of these files were not executable, then we would have backed-off to a system-wide installation (as the Conda installation check would not have been validated)
- now, if either one of these files are not executable, then we would raise an exception without even verifying whether a system-wide installation is available or not.

Hence, if we want to keep the new behavior, I would document the change in a comment and/or in the commit message.

**Review by popescu-v — CHANGES_REQUESTED** [review]
See the replies to the comments.

File: khiops/core/api.py [hunk 4]
```diff
@@ -174,17 +174,39 @@ def _preprocess_task_arguments(task_args):
...
     # Set the discretization/grouping default values
     if "discretization_method" in task_args:
+        # Set the default values if the discretization method is not specified
         if task_args["discretization_method"] is None:
             if task_args["target_variable"]:
                 task_args["discretization_method"] = "MODL"
             else:
                 task_args["discretization_method"] = "EqualWidth"
+        # Otherwise raise an error if the values are not in the range to avoid a khiops
+        # segmentation fault. This won't be necessary from version 11 on.
```
Comment (popescu-v) [inline khiops/core/api.py:184]: Could you add a `# TODO: Remove in Khiops 11` so that we do not forget about this?

File: khiops/core/api.py [hunk 5]
```diff
@@ -174,17 +174,39 @@ def _preprocess_task_arguments(task_args):
...
+
     if "grouping_method" in task_args:
+        # Set the default values if the grouping method is not specified
         if task_args["grouping_method"] is None:
             if task_args["target_variable"]:
                 task_args["grouping_method"] = "MODL"
             else:
                 task_args["grouping_method"] = "BasicGrouping"
+        # Otherwise raise an error if the values are not in the range to avoid a khiops
+        # segmentation fault. This won't be necessary from version 11 on.
```
Comment (popescu-v) [inline khiops/core/api.py:202]: See above (add `# TODO: Remove in Khiops 11`)?

File: khiops/core/internals/runner.py [hunk 6]
```diff
@@ -616,51 +618,77 @@ def run(
...
-        """
-        # If there were no errors warn if:
+    def _report_exit_status(
+        self, tool_name, return_code, stdout, stderr, log_file_path
+    ):
+        """Reports the exit status of a Khiops execution"""
+        # Note:
+        #   We report stdout and stderr below because Khiops normally these streams
+        #   should be empty because we always use the log file. In practice MPI and the
+        #   remote filesystems plugins may write to these streams.
```
Comment (popescu-v) [inline khiops/core/internals/runner.py:628]: I would slightly rephrase to something like:
```python
# Normally, Khiops leaves the stdout and stderr streams
# empty, as we always use the log file. In practice, MPI and the 
# remote filesystem plugins may write to these streams.
# Hence, we report them.
```

File: khiops/core/internals/runner.py (hunk 3 above)
Comment (popescu-v) [reply khiops/core/internals/runner.py:661]: Ok, pending the slight rephrasal (see comment above).

**Review by popescu-v — CHANGES_REQUESTED** [review]
See all pending comments.

File: khiops/core/internals/runner.py (hunk 1 above)
Comment (folmos-at-orange) [reply khiops/core/internals/runner.py:179]: I'll try to refactor` _check_tools` to recover the old behavior.

File: khiops/core/internals/runner.py (hunk 3 above)
Comment (folmos-at-orange) [reply khiops/core/internals/runner.py:661]: It is done and explained in the note I made. I don't get what is missing.

File: khiops/core/api.py (hunk 4 above)
Comment (folmos-at-orange) [reply khiops/core/api.py:184]: I prefer make an issue.

File: khiops/core/api.py (hunk 5 above)
Comment (folmos-at-orange) [reply khiops/core/api.py:202]: Se above.

File: khiops/core/internals/runner.py (hunk 1 above)
Comment (popescu-v) [reply khiops/core/internals/runner.py:179]: Ok

File: khiops/core/internals/runner.py (hunk 6 above)
Comment (folmos-at-orange) [reply khiops/core/internals/runner.py:628]: I prefer active phrasing because with passive the most important stuff is at the end.  I'll rephrase in the spirit of your comment but keep the active phrasing.

File: khiops/core/internals/runner.py (hunk 3 above)
Comment (popescu-v) [reply khiops/core/internals/runner.py:661]: I meant that I would rephrase the currently written text (
```
#   We report stdout and stderr below because Khiops normally these streams
#   should be empty because we always use the log file. In practice MPI and the
#   remote filesystems plugins may write to these streams.
```
)
to

```
# Normally, Khiops leaves the stdout and stderr streams
# empty, as we always use the log file. In practice, MPI and the 
# remote filesystem plugins may write to these streams.
# Hence, we report them.
```

**Comment by popescu-v** [comment]
> > > > I do not see the relation between the name of this PR and the implemented changes. IMHO, we need to provide description and context (ticket) to this PR, or otherwise rename the PR to reflect the implemented changes.
> > > 
> > > 
> > > Just minor fixes that do not require another pull request. To go fast.
> > 
> > 
> > I would just rename the PR.
> 
> I would not bother with it. What's important are the commits. The PR may dissapear tomorrow.

Ok with me (I cannot close this).

File: khiops/core/internals/runner.py (hunk 6 above)
Comment (popescu-v) [reply khiops/core/internals/runner.py:628]: Ok

File: khiops/extras/docker.py (hunk 2 above)
Comment (folmos-at-orange) [reply khiops/extras/docker.py:165]: Ok

**Comment by folmos-at-orange** [comment]
> > > > > I do not see the relation between the name of this PR and the implemented changes. IMHO, we need to provide description and context (ticket) to this PR, or otherwise rename the PR to reflect the implemented changes.
> > > > 
> > > > 
> > > > Just minor fixes that do not require another pull request. To go fast.
> > > 
> > > 
> > > I would just rename the PR.
> > 
> > 
> > I would not bother with it. What's important are the commits. The PR may dissapear tomorrow.
> 
> Ok (I am not able to close this thread).

This is the main PR thread and cannot be "closed". Maybe we should always comment in the code to have sub-threads (for example at the beginning of a file and putting in **General Comment** at the beginning).

File: khiops/extras/docker.py (hunk 2 above)
Comment (folmos-at-orange) [reply khiops/extras/docker.py:165]: Done

File: khiops/core/internals/runner.py (hunk 1 above)
Comment (folmos-at-orange) [reply khiops/core/internals/runner.py:179]: Thinking a little bit, if the executables exists in the conda bin directory and they are not executable we should no try to find a system-wide installation because it is probably an installation problem.

Comment (folmos-at-orange) [reply khiops/core/internals/runner.py:179]: I refactored, but keep the logic that when the `MODL*` files are found in the conda env but they are not executable it is an error (reported by `_check_tools`).

Comment (popescu-v) [reply khiops/core/internals/runner.py]: Ok, this would help the see that there was a Conda installation problem.

**Review by popescu-v — APPROVED** [review]
LGTM.
